### PR TITLE
Relocate coroutine upvars into Unresumed state

### PR DESCRIPTION
**Pull request by dingxiangfei2009**
Related to #62958

This PR is an attempt to address the async/coroutine size issue by allowing independent def-use/liveness analysis on individual upvars in coroutines. It has appeared to address partially the size doubling issue introduced by the use of upvars.

However, there are caveats detailed in the following list that I would like to address before turning this draft in.
- The treatment here towards the `ty::Coroutine` in MIR passes is unfortunately "messier" than my liking, which is something I definitely want to change. I propose to promote upvars into `Body<'tcx>` along with `local_decls`, so that we can safely handle them safely. I would happily open a new separate PR to improve the upvar management.
- It is not a generic solution, yet. For instance, we are still doubling the size in the example of #62958. If we insert a pass before MIR type analysis to remove unnecessary drops, which we can, that particular size doubling will be solved. However, if a `Future` upvar is alive across more than one yield points, that upvar is still ineligible. It makes sense because we would like to minimize moving of variant fields. How to handle these upvars is not the focus of this PR for now.

Out of expectation of possible change in the high level plan, I am keeping this as a draft in hope of invoking conversations. :bow: 

cc @pnkfelix for the context.

### CODE REVIEW

**Comment by pnkfelix** [comment]
@dingxiangfei2009 am I right in assuming that the debugger test failure is expected and you plan to address it if we move forward with this approach?

File: compiler/rustc_codegen_cranelift/src/base.rs [hunk 1]
```diff
@@ -780,6 +780,9 @@ fn codegen_stmt<'tcx>(
                             let variant_dest = lval.downcast_variant(fx, variant_index);
                             (variant_index, variant_dest, active_field_index)
                         }
+                        mir::AggregateKind::Coroutine(_, _) => {
+                            (FIRST_VARIANT, lval.downcast_variant(fx, FIRST_VARIANT), None)
```
Comment (pnkfelix) [inline compiler/rustc_codegen_cranelift/src/base.rs:784]: The fact that the unresumed state is the first variant, that's just an assumption that this code is making, right?

I'm not certain how heavily that assumption is baked into all the rest of the code. I do see that the variant `CoroutineArgs::UNRESUMED` is assigned the value 0, but I don't know if that value is set in stone.

(Maybe all I'm looking for is an assertion ensuring `FIRST_VARIANT` *is* the variant for Unresumed on all these places where you downcast to FIRST_VARIANT.)

**Comment by pnkfelix** [comment]
In any case, this is interesting work, and I *think* its a step in the right direction.

I plan to build it locally and take it for a spin, to make sure I understand what its doing. @dingxiangfei2009, I may ping you on Zulip with follow-up questions.

**Comment by dingxiangfei2009** [comment]
@pnkfelix Thank you for reviewing! Yes, the debuginfo test may very much depend on how the upvars are eventually represented. I figured that I shall fix them after we could settle with our approach first.

Please feel free to message on Zulip!

File: compiler/rustc_codegen_cranelift/src/base.rs (hunk 1 above)
Comment (dingxiangfei2009) [reply compiler/rustc_codegen_cranelift/src/base.rs:784]: That's correct. `UNRESUMED` is used across, except in the codegen backends which is `FIRST_VARIANT`. It appears that the first variant is assumed for coroutines and `struct`s, though this fact was not obvious at the first glance.

File: compiler/rustc_mir_transform/src/coroutine.rs [hunk 2]
```diff
@@ -1105,11 +1168,25 @@ fn compute_layout<'tcx>(
             // around inside coroutines, so it doesn't matter which variant
```
Comment (pnkfelix) [inline compiler/rustc_mir_transform/src/coroutine.rs:1168]: This comment was cut off in the patch, it starts off with:
```
             // Note that if a field is included in multiple variants, we will
            // just use the first one here. That's fine; fields do not move
```

I had started writing this PR comment with a question asking if this code comment needs revision for consistency with this change. After reflecting on things, i *think* the whole picture here remains coherent, but now I want to double check my interpretation.

Before this patch, the upvars were disjoint from the fields carried across yields of a generator/coroutine, so the comment above simply didn't apply to upvars.

After this patch, the upvars are now mapped into the Unresumed variant of a generator/coroutine. But that isn't inconsistent with the comment above ... as long as the Unresumed variant is itself the *first* variant... right?

File: compiler/rustc_mir_transform/src/coroutine.rs [hunk 3]
```diff
@@ -1221,6 +1298,21 @@ fn elaborate_coroutine_drops<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
         );
     }
     elaborator.patch.apply(body);
+    for block_data in body.basic_blocks_mut() {
```
Comment (pnkfelix) [inline compiler/rustc_mir_transform/src/coroutine.rs:1301]: Is there some code up above this (or called by one of the methods above) that was in principle responsible for dropping the upvars, that is now made obsolete under the new regime you have introduced here?

**Comment by pnkfelix** [comment]
Thanks for your patience @dingxiangfei2009 . This all looks good to me; I wouldn't mind getting an answer to [my question about the new handling of dropping upvar state](https://github.com/rust-lang/rust/pull/120168#discussion_r1509371355), but I also think this is ready to land, assuming it doesn't conflict with @compiler-errors 's ongoing work on async closures etc.

**Comment by pnkfelix** [comment]
Oh, also: Just in case its potentially of interest to others, part of the reason this took me so long to get to is that I wanted a pair of pernos.co runs, one from before and one from after this PR, to make it easier for me to explore the control flow interactively.

Here's links to those in case anyone else wants to inspect them:

 * Before: https://pernos.co/debug/JHlinh-cWSwzHKGoZRiTeQ/index.html
 * After: https://pernos.co/debug/SEXi12zlbIZomhOhyuoegg/index.html

**Comment by dingxiangfei2009** [comment]
Progress update: I am still on it to polish it up. I will try to get the rust-timer numbers this week.

File: compiler/rustc_const_eval/src/transform/validate.rs [hunk 4]
```diff
@@ -109,14 +109,17 @@ impl<'tcx> MirPass<'tcx> for Validator {
...
                 // computation code.
-                cfg_checker.fail(
-                    Location::START,
-                    format!(
-                        "Coroutine layout differs from by-move coroutine layout:\n\
-                        layout: {layout:#?}\n\
-                        by_move_layout: {by_move_layout:#?}",
-                    ),
-                );
+                // NOTE(@dingxiangfei2009): However, given that upvars are now part of the layout
```
Comment (compiler-errors) [inline compiler/rustc_const_eval/src/transform/validate.rs:112]: This is not correct, and will cause codegen problems.

Comment (compiler-errors) [reply compiler/rustc_const_eval/src/transform/validate.rs:112]: The comment says "If this turns out not to be true, please let compiler-errors know.", but you just commented it out 😓 Please make sure this does not land in this state.

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/rustc_const_eval/src/transform/validate.rs (hunk 4 above)
Comment (dingxiangfei2009) [reply compiler/rustc_const_eval/src/transform/validate.rs:112]: I will add the appropriate tests here.

The next thing to do is to check the the byref captures and make sure that they do match up modulo references, and the other parts of the layout are the same.

So far, by checking the codegen results, I think the outputs look right but I will be extra careful! :pray:

Comment (compiler-errors) [reply compiler/rustc_const_eval/src/transform/validate.rs:112]: What do you mean by "they do match up modulo references"?

The problem here is that when we call `tcx.coroutine_layout()` it is not going to be correct for by-move async closures that capture references.

Comment (dingxiangfei2009) [reply compiler/rustc_const_eval/src/transform/validate.rs:112]: Yes, that is going to be a problem. I think I should change the coroutine_layout query to accept the closure kind, too. You can see that in other places in this patch, I have to check the correct kind type to determine if the by-move layout or the non-by-move layout shall be used. I find it error-prone so I am going to change thatq.

Comment (compiler-errors) [reply compiler/rustc_const_eval/src/transform/validate.rs:112]: I believe that's the correct change, however I've already started to make that change (starting with https://github.com/rust-lang/rust/pull/123021) so I'll have a PR out in a few minutes...

File: src/tools/clippy/tests/ui/crashes/ice-10645.stderr [hunk 5]
```diff
@@ -4,11 +4,14 @@ warning: future cannot be sent between threads safely
...
-  --> tests/ui/crashes/ice-10645.rs:5:29
+note: future is not `Send` as this value is used across an await
+  --> tests/ui/crashes/ice-10645.rs:5:35
    |
 LL | pub async fn bar<'a, T: 'a>(_: T) {}
-   |                             ^ has type `T` which is not `Send`
+   |                                   ^-
+   |                                   |
+   |                                   await occurs here, with `{}` maybe used later
+   |                                   has type `T` which is not `Send`
```
Comment (flip1995) [inline src/tools/clippy/tests/ui/crashes/ice-10645.stderr:14]: The Clippy changes do not look correct. The new `note` now claims, that the body of the function is maybe used later.  You probably have to fix the Clippy code, rather than just update the stderr files.

**Review by flip1995 — CHANGES_REQUESTED** [review]
Clippy fallout should block this PR until addressed.

File: src/tools/clippy/tests/ui/crashes/ice-10645.stderr (hunk 5 above)
Comment (dingxiangfei2009) [reply src/tools/clippy/tests/ui/crashes/ice-10645.stderr:14]: I see. I will work on restoring the identification on the upvar instead of erring on the `UNRESUME` variant.

Comment (dingxiangfei2009) [reply src/tools/clippy/tests/ui/crashes/ice-10645.stderr]: I would like to report that the span is now corrected.

File: compiler/rustc_const_eval/src/transform/validate.rs (hunk 4 above)
Comment (dingxiangfei2009) [reply compiler/rustc_const_eval/src/transform/validate.rs]: Thank you for the PR! I have rebased onto the work and it looks much cleaner! :bow:

File: src/tools/clippy/tests/ui/large_futures.rs [hunk 6]
```diff
@@ -33,7 +33,6 @@ pub async fn test() {
     foo().await;
     //~^ ERROR: large future with a size of 65540 bytes
     calls_fut(fut).await;
-    //~^ ERROR: large future with a size of 49159 bytes
```
Comment (flip1995) [inline src/tools/clippy/tests/ui/large_futures.rs:36]: Please don't remove this annotation, but fix the number to be correct: `32774`.

File: src/tools/clippy/tests/ui/needless_lifetimes.rs [hunk 7]
```diff
@@ -527,7 +527,7 @@ mod in_macro {
 }
 
 mod issue5787 {
-    use std::sync::MutexGuard;
+    struct MutexGuard<'a, T>(std::marker::PhantomData<fn() -> &'a T>);
```
Comment (flip1995) [inline src/tools/clippy/tests/ui/needless_lifetimes.rs:530]: What's the motivation of this change?

File: src/tools/clippy/tests/ui/crashes/ice-10645.stderr (hunk 5 above)
Comment (flip1995) [reply src/tools/clippy/tests/ui/crashes/ice-10645.stderr]: Thanks!

File: src/tools/clippy/tests/ui/needless_lifetimes.rs (hunk 7 above)
Comment (dingxiangfei2009) [reply src/tools/clippy/tests/ui/needless_lifetimes.rs:530]: Thanks for the reminder! rust-lang/rust-clippy#5787 is about unnecessary lint on the `'a` lifetime in the `Foo::wait` signature below. By putting `MutexGuard` onto the coroutine variant instead, clippy now lints loudly that `MutexGuard` is held across `await`. This is technically not related to issue 5787.

However, I think I should add a test in `src/tools/clippy/tests/ui/await_holding_lock.rs` for this case, because holding a `MutexGuard` from `std` in an upvar is equally bad and shall be linted. Even if the guard is dropped at the beginning of the async body, it will be a problem if the future is instantiated but not polled. I am working on this clippy test.

Comment (flip1995) [reply src/tools/clippy/tests/ui/needless_lifetimes.rs:530]: I see. In this case we usually just allow the second lint that triggers unexpectedly. I would prefer if you add an `allow(await_holding_lock)` here, over re-defining `MutexGuard`. That would be way easier to understand and explain to a reader of this test case.

Comment (dingxiangfei2009) [reply src/tools/clippy/tests/ui/needless_lifetimes.rs:530]: The new test is in.

File: src/tools/clippy/tests/ui/await_holding_lock.stderr [hunk 8]
```diff
@@ -55,80 +55,97 @@ LL |
...
+LL | |
+LL | |         drop(x);
+LL | |     }
+   | |_____^
+   |
+   = help: consider using an async-aware `Mutex` type or ensuring the `MutexGuard` is dropped before calling await
+note: these are all the `await` points this lock is held through
+  --> tests/ui/await_holding_lock.rs:64:54
+   |
+LL |     pub async fn equally_bad(x: MutexGuard<'_, u32>) {
```
[the remark this answers is not in the document]
Comment (dingxiangfei2009) [reply src/tools/clippy/tests/ui/await_holding_lock.stderr:72]: The wording `these are all the `await` points` is kind of shared across all the suspension point. The very first suspension point is, yes, the construction of the async state machine/coroutine. You may think of it as the implicit first `.await` point.

[the remark this answers is not in the document]
Comment (dingxiangfei2009) [reply src/tools/clippy/tests/ui/await_holding_lock.stderr:72]: Forgive me for tagging for reviewing again without a new change-set.

The lint triggers in this case because a lock guard is assigned into the coroutine/async body. I personally think the lint message here makes sense because the idea is to deny the passing of non-`async` `MutexGuard`s into `async` bodies.

I think the wording needs some adjustment.

[the remark this answers is not in the document]
Comment (dingxiangfei2009) [reply src/tools/clippy/tests/ui/await_holding_lock.stderr:72]: Wording adjusted. Another clippy test is added to show that it applies to `std::cell::Ref` as well.

[the remark this answers is not in the document]
Comment (dingxiangfei2009) [reply src/tools/clippy/tests/ui/await_holding_lock.stderr:72]: I am figuring out how to extract span for the upvars. One moment.

File: tests/ui/type-alias-impl-trait/indirect-recursion-issue-112047.stderr [hunk 9]
```diff
@@ -6,13 +6,6 @@ LL |         async move { recur(self).await; }
    |                      |
    |                      recursive call here
    |
-note: which leads to this async fn
```
Comment (dingxiangfei2009) [inline tests/ui/type-alias-impl-trait/indirect-recursion-issue-112047.stderr:9]: I will restore this note later.

File: src/tools/clippy/tests/ui/await_holding_lock.stderr (hunk 8 above)
[the remark this answers is not in the document]
Comment (flip1995) [reply src/tools/clippy/tests/ui/await_holding_lock.stderr:68]: Now the note in this case is a bit pointless. The error message says "this argument is passed" and points to the argument and the note afterwards says the exact same thing again with a different span. Could you remove the note in this case?

**Review by flip1995 — COMMENTED** [review]
One more comment about the Clippy changes. I generally like the new suggestion building for this lint, but I think it can be simplified a bit.

File: tests/ui/type-alias-impl-trait/indirect-recursion-issue-112047.stderr (hunk 9 above)
Comment (dingxiangfei2009) [reply tests/ui/type-alias-impl-trait/indirect-recursion-issue-112047.stderr:9]: Okay I am figuring out why the span is missing from the query frame. :thinking:

**Comment by dingxiangfei2009** [comment]
@flip1995 Somehow I can't respond to your change request. I have decided to skip over the upvars in the additional notes.

**Review by flip1995 — APPROVED** [review]
Awesome! Clippy changes approved, so no longer blocking this PR from the Clippy side.

@dingxiangfei2009 The not-respond issue is a GitHub weirdness: If you respond to an open thread with a review comment, the comment will be posted twice: Once on the original thread where you can respond and once in the review where you can't respond.

File: compiler/rustc_mir_dataflow/src/framework/engine.rs [hunk 10]
```diff
@@ -288,7 +288,7 @@ where
             }
 
             None if dump_enabled(tcx, A::NAME, def_id) => {
-                create_dump_file(tcx, ".dot", false, A::NAME, &pass_name.unwrap_or("-----"), body)?
+                create_dump_file(tcx, ".dot", true, A::NAME, &pass_name.unwrap_or("-----"), body)?
```
Comment (pnkfelix) [inline compiler/rustc_mir_dataflow/src/framework/engine.rs:291]: I don't object to this change, but it isn't strictly speaking necessary for this PR, is it?

File: compiler/rustc_mir_dataflow/src/impls/liveness.rs [hunk 11]
```diff
@@ -74,22 +131,30 @@ impl<'tcx> GenKillAnalysis<'tcx> for MaybeLiveLocals {
...
                 }
             });
         }
     }
 }
 
-pub struct TransferFunction<'a, T>(pub &'a mut T);
+pub struct TransferFunction<'a, T> {
+    pub trans: &'a mut T,
+    pub upvars: Option<(Local, usize)>,
```
Comment (pnkfelix) [inline compiler/rustc_mir_dataflow/src/impls/liveness.rs:156]: I don't remember if earlier iterations on this PR had these changes to the various dataflow impls; its possible they did and I just overlooked them.

But seeing them now with fresh eyes, it just seems unfortunate that so many dataflow implementations seem like they are expected to duplicate this `upvars: Option<(Local, usize)>` state pattern. I would be very concerned about whether it is a footgun to forget to do it, and also whether it makes all of the dataflow systems harder to maintain.

Having said that, I also don't have a better suggestion about what to do for the short term. It probably is something to bring up with the broader team, in terms of explaining why the pattern is necessary and seeing if anyone has ideas for another way to accomplish the objective.

**Comment by pnkfelix** [comment]
I'm going to nominate this PR for discussion at next week's T-compiler meeting.

The essential question I want to pose is: Are we okay with the `upvars: Option<(Local, usize)>` this change is adding to various dataflow analyses, at least in the short term so that we can make forward progress on the problem of generator size explosion ?

Or should we have a discussion between @dingxiangfei2009, @pnkfelix, and whichever team members are interested in any of {dataflow, MIR representation, generator representation} to try to tease how what the "right answer" is here.

@rustbot label: +I-compiler-nominated

**Comment by pnkfelix** [comment]
Okay, the team discussed briefly at our team meeting.

@dingxiangfei2009 has opened up a dedicated zulip topic to discuss the matter, at: https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Seeking.20opinions.20about.20lifting.20upvars.20in.20coroutine.20and.20fu.2E.2E.2E

In the above zulip topic, @dingxiangfei2009 wrote:

> I am still investigating if we can split the BitSet domain into Locals and Upvars. Despite the extensive exposure to assumptions that only true Locals exists, I believe that it is still doable, so that existing and new analyses can be sound again, whether they need to handle upvars or not.

In the [team discussion](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-04-18/near/434092526), we generally agreed that we'd prefer to see @dingxiangfei2009 explore that split-bitset alternative design, because the approach in this PR seems more brittle than we would like, assuming there are better alternatives.

@rustbot author

**Comment by oskgo** [comment]
Triage: Closing since this is replaced by #127522.

Note: if you are going to continue please open the PR BEFORE you push to it, else you won't be able to reopen - this is a quirk of github.
Thanks for your contribution.

@rustbot label: +S-inactive
